File: java/src/test/java/org/libpetri/core/internal/SubnetRewriterTest.java
package org.libpetri.core.internal;

import org.junit.jupiter.api.Test;
import org.libpetri.core.Arc;
import org.libpetri.core.PetriNet;
import org.libpetri.core.Place;
import org.libpetri.core.Transition;

import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for {@link SubnetRewriter} — the structural rewrite primitive
 * shared between {@code SubnetDef.instantiate(...)} and (later)
 * {@code PetriNet.Builder.compose(...)} per <b>MOD-020</b>.
 *
 * <p>Coverage:
 * <ul>
 *   <li>Exhaustive {@code Arc.In} variants (One, Exactly, All, AtLeast).</li>
 *   <li>Exhaustive {@code Arc.Out} variants (Place, And, Xor, Timeout, ForwardInput),
 *       including a recursive {@code And/Xor} tree to verify tree-walking.</li>
 *   <li>Coverage of {@code Inhibitor}, {@code Read}, {@code Reset}.</li>
 *   <li>Identity behaviour on places not present in the remap (partial-remap
 *       semantics — the future {@code compose} caller depends on this).</li>
 *   <li>{@code renameNet} end-to-end check on a small body.</li>
 * </ul>
 */
class SubnetRewriterTest {

    // ============================================================
    //  Helpers
    // ============================================================

    private static <T> Place<T> p(String name, Class<T> type) {
        return Place.of(name, type);
    }

    /** Builds a single-entry remap: orig -> renamed (same tokenType). */
    private static <T> Map<Place<?>, Place<?>> remap(Place<T> orig, String prefix) {
        var m = new HashMap<Place<?>, Place<?>>();
        m.put(orig, SubnetRewriter.renamePlace(orig, prefix));
        return m;
    }

    private static Place<?> renamed(Map<Place<?>, Place<?>> m, Place<?> orig) {
        return m.get(orig);
    }

    // ============================================================
    //  Arc.In rewrite — exhaustive variant coverage
    // ============================================================

    @Test
    void rewriteIn_one_remapsPlace() {
        Place<String> orig = p("input", String.class);
        var m = remap(orig, "r1");

        var rewritten = SubnetRewriter.rewriteIn(new Arc.In.One(orig), m);

        assertInstanceOf(Arc.In.One.class, rewritten);
        assertEquals("r1/input", rewritten.place().name());
        assertEquals(String.class, rewritten.place().tokenType());
    }

    @Test
    void rewriteIn_exactly_remapsPlace_preservesCount() {
        Place<String> orig = p("batch", String.class);
        var m = remap(orig, "r1");

        var rewritten = SubnetRewriter.rewriteIn(new Arc.In.Exactly(orig, 7), m);

        assertInstanceOf(Arc.In.Exactly.class, rewritten);
        var ex = (Arc.In.Exactly) rewritten;
        assertEquals("r1/batch", ex.place().name());
        assertEquals(7, ex.count());
    }

    @Test
    void rewriteIn_all_remapsPlace() {
        Place<String> orig = p("queue", String.class);
        var m = remap(orig, "r1");

        var rewritten = SubnetRewriter.rewriteIn(new Arc.In.All(orig), m);

        assertInstanceOf(Arc.In.All.class, rewritten);
        assertEquals("r1/queue", rewritten.place().name());
    }

    @Test
    void rewriteIn_atLeast_remapsPlace_preservesMinimum() {
        Place<String> orig = p("buf", String.class);
        var m = remap(orig, "r1");

        var rewritten = SubnetRewriter.rewriteIn(new Arc.In.AtLeast(orig, 3), m);

        assertInstanceOf(Arc.In.AtLeast.class, rewritten);
        var al = (Arc.In.AtLeast) rewritten;
        assertEquals("r1/buf", al.place().name());
        assertEquals(3, al.minimum());
    }

    // ============================================================
    //  Arc.Out rewrite — exhaustive variant coverage + recursion
    // ============================================================

    @Test
    void rewriteOut_place_remapsPlace() {
        Place<String> orig = p("out", String.class);
        var m = remap(orig, "r1");

        var rewritten = SubnetRewriter.rewriteOut(new Arc.Out.Place(orig), m);

        assertInstanceOf(Arc.Out.Place.class, rewritten);
        assertEquals("r1/out", ((Arc.Out.Place) rewritten).place().name());
    }

    @Test
    void rewriteOut_and_remapsAllChildren() {
        Place<String> a = p("a", String.class);
        Place<String> b = p("b", String.class);
        var m = new HashMap<Place<?>, Place<?>>();
        m.put(a, SubnetRewriter.renamePlace(a, "r1"));
        m.put(b, SubnetRewriter.renamePlace(b, "r1"));

        var and = new Arc.Out.And(List.of(new Arc.Out.Place(a), new Arc.Out.Place(b)));
        var rewritten = (Arc.Out.And) SubnetRewriter.rewriteOut(and, m);

        assertEquals(2, rewritten.children().size());
        var c0 = (Arc.Out.Place) rewritten.children().get(0);
        var c1 = (Arc.Out.Place) rewritten.children().get(1);
        assertEquals("r1/a", c0.place().name());
        assertEquals("r1/b", c1.place().name());
    }

    @Test
    void rewriteOut_xor_remapsAllChildren() {
        Place<String> a = p("a", String.class);
        Place<String> b = p("b", String.class);
        var m = new HashMap<Place<?>, Place<?>>();
        m.put(a, SubnetRewriter.renamePlace(a, "r1"));
        m.put(b, SubnetRewriter.renamePlace(b, "r1"));

        var xor = new Arc.Out.Xor(List.of(new Arc.Out.Place(a), new Arc.Out.Place(b)));
        var rewritten = (Arc.Out.Xor) SubnetRewriter.rewriteOut(xor, m);

        assertEquals(2, rewritten.children().size());
        assertEquals("r1/a", ((Arc.Out.Place) rewritten.children().get(0)).place().name());
        assertEquals("r1/b", ((Arc.Out.Place) rewritten.children().get(1)).place().name());
    }

    @Test
    void rewriteOut_andOfXor_recursivelyRewrites() {
        Place<String> a = p("a", String.class);
        Place<String> b = p("b", String.class);
        Place<String> c = p("c", String.class);
        var m = new HashMap<Place<?>, Place<?>>();
        m.put(a, SubnetRewriter.renamePlace(a, "r1"));
        m.put(b, SubnetRewriter.renamePlace(b, "r1"));
        m.put(c, SubnetRewriter.renamePlace(c, "r1"));

        // AND( XOR(a,b), Place(c) )
        var tree = new Arc.Out.And(List.of(
            new Arc.Out.Xor(List.of(new Arc.Out.Place(a), new Arc.Out.Place(b))),
            new Arc.Out.Place(c)
        ));

        var rewritten = (Arc.Out.And) SubnetRewriter.rewriteOut(tree, m);

        // Outer And shape preserved.
        assertEquals(2, rewritten.children().size());

        // First child must still be an Xor, with both leaves remapped.
        var innerXor = (Arc.Out.Xor) rewritten.children().get(0);
        assertEquals(2, innerXor.children().size());
        assertEquals("r1/a", ((Arc.Out.Place) innerXor.children().get(0)).place().name());
        assertEquals("r1/b", ((Arc.Out.Place) innerXor.children().get(1)).place().name());

        // Second child is a leaf Place, remapped.
        assertEquals("r1/c", ((Arc.Out.Place) rewritten.children().get(1)).place().name());
    }

    @Test
    void rewriteOut_xorOfAnd_recursivelyRewrites() {
        Place<String> a = p("a", String.class);
        Place<String> b = p("b", String.class);
        Place<String> c = p("c", String.class);
        Place<String> d = p("d", String.class);
        var m = new HashMap<Place<?>, Place<?>>();
        m.put(a, SubnetRewriter.renamePlace(a, "r1"));
        m.put(b, SubnetRewriter.renamePlace(b, "r1"));
        m.put(c, SubnetRewriter.renamePlace(c, "r1"));
        m.put(d, SubnetRewriter.renamePlace(d, "r1"));

        // XOR( AND(a,b), AND(c,d) )
        var tree = new Arc.Out.Xor(List.of(
            new Arc.Out.And(List.of(new Arc.Out.Place(a), new Arc.Out.Place(b))),
            new Arc.Out.And(List.of(new Arc.Out.Place(c), new Arc.Out.Place(d)))
        ));

        var rewritten = (Arc.Out.Xor) SubnetRewriter.rewriteOut(tree, m);
        assertEquals(2, rewritten.children().size());

        var leftAnd  = (Arc.Out.And) rewritten.children().get(0);
        var rightAnd = (Arc.Out.And) rewritten.children().get(1);

        assertEquals("r1/a", ((Arc.Out.Place) leftAnd.children().get(0)).place().name());
        assertEquals("r1/b", ((Arc.Out.Place) leftAnd.children().get(1)).place().name());
        assertEquals("r1/c", ((Arc.Out.Place) rightAnd.children().get(0)).place().name());
        assertEquals("r1/d", ((Arc.Out.Place) rightAnd.children().get(1)).place().name());
    }

    @Test
    void rewriteOut_timeout_recursesIntoChild_preservesDuration() {
        Place<String> orig = p("late", String.class);
        var m = remap(orig, "r1");

        var t = new Arc.Out.Timeout(Duration.ofMillis(250), new Arc.Out.Place(orig));
        var rewritten = (Arc.Out.Timeout) SubnetRewriter.rewriteOut(t, m);

        assertEquals(Duration.ofMillis(250), rewritten.after());
        assertEquals("r1/late", ((Arc.Out.Place) rewritten.child()).place().name());
    }

    @Test
    void rewriteOut_forwardInput_remapsBothEnds() {
        Place<String> from = p("query", String.class);
        Place<String> to   = p("retry", String.class);
        var m = new HashMap<Place<?>, Place<?>>();
        m.put(from, SubnetRewriter.renamePlace(from, "r1"));
        m.put(to,   SubnetRewriter.renamePlace(to, "r1"));

        var fi = new Arc.Out.ForwardInput(from, to);
        var rewritten = (Arc.Out.ForwardInput) SubnetRewriter.rewriteOut(fi, m);

        assertEquals("r1/query", rewritten.from().name());
        assertEquals("r1/retry", rewritten.to().name());
    }

    // ============================================================
    //  Inhibitor / Read / Reset
    // ============================================================

    @Test
    void rewriteInhibitor_remapsPlace_preservesGenericType() {
        Place<String> orig = p("paused", String.class);
        var m = remap(orig, "r1");

        var rewritten = SubnetRewriter.rewriteInhibitor(new Arc.Inhibitor<>(orig), m);
        assertEquals("r1/paused", rewritten.place().name());
        assertEquals(String.class, rewritten.place().tokenType());
    }

    @Test
    void rewriteRead_remapsPlace() {
        Place<String> orig = p("ctx", String.class);
        var m = remap(orig, "r1");

        var rewritten = SubnetRewriter.rewriteRead(new Arc.Read<>(orig), m);
        assertEquals("r1/ctx", rewritten.place().name());
    }

    @Test
    void rewriteReset_remapsPlace() {
        Place<String> orig = p("buf", String.class);
        var m = remap(orig, "r1");

        var rewritten = SubnetRewriter.rewriteReset(new Arc.Reset<>(orig), m);
        assertEquals("r1/buf", rewritten.place().name());
    }

    // ============================================================
    //  Partial remap (identity on unmapped places)
    // ============================================================

    @Test
    void rewriteIn_partialRemap_unmappedPlaceIsIdentity() {
        // Critical for compose(...): the future caller will pass a partial
        // remap (only the bound port places). Unmapped places must stay put.
        Place<String> mapped   = p("mapped", String.class);
        Place<String> unmapped = p("unmapped", String.class);
        var m = remap(mapped, "r1");
        // 'unmapped' is intentionally absent from m.

        var rA = SubnetRewriter.rewriteIn(new Arc.In.One(mapped), m);
        assertEquals("r1/mapped", rA.place().name(), "mapped place is rewritten");

        var rB = SubnetRewriter.rewriteIn(new Arc.In.One(unmapped), m);
        assertSame(unmapped, rB.place(), "unmapped place is identity (not renamed)");
    }

    @Test
    void rewriteOut_partialRemap_unmappedLeafIsIdentity() {
        Place<String> mapped   = p("mapped", String.class);
        Place<String> unmapped = p("unmapped", String.class);
        var m = remap(mapped, "r1");

        var tree = new Arc.Out.And(List.of(
            new Arc.Out.Place(mapped),
            new Arc.Out.Place(unmapped)
        ));
        var rewritten = (Arc.Out.And) SubnetRewriter.rewriteOut(tree, m);
        assertEquals("r1/mapped", ((Arc.Out.Place) rewritten.children().get(0)).place().name());
        assertSame(unmapped, ((Arc.Out.Place) rewritten.children().get(1)).place());
    }

    // ============================================================
    //  renameNet end-to-end
    // ============================================================

    @Test
    void renameNet_endToEnd_renamesEverythingAndFillsMaps() {
        Place<String> in = p("in", String.class);
        Place<String> out = p("out", String.class);
        Transition t = Transition.builder("forward")
            .inputs(Arc.In.one(in))
            .outputs(Arc.Out.and(out))
            .build();
        PetriNet body = PetriNet.builder("Body").transition(t).build();

        var placeRemap = new HashMap<Place<?>, Place<?>>();
        var transitionRemap = new HashMap<Transition, Transition>();

        PetriNet renamed = SubnetRewriter.renameNet(body, "i1", placeRemap, transitionRemap);

        // Net name itself prefixed.
        assertEquals("i1/Body", renamed.name());

        // Every place renamed.
        assertEquals(2, renamed.places().size());
        for (var p : renamed.places()) {
            assertTrue(p.name().startsWith("i1/"), "place: " + p.name());
        }

        // The remap holds old → new bindings.
        assertEquals("i1/in",  placeRemap.get(in).name());
        assertEquals("i1/out", placeRemap.get(out).name());

        // The single transition is also renamed and recorded in the remap.
        assertEquals(1, renamed.transitions().size());
        var renamedT = renamed.transitions().iterator().next();
        assertEquals("i1/forward", renamedT.name());
        assertSame(renamedT, transitionRemap.get(t));

        // Arcs in the renamed transition reference the renamed places (not originals).
        assertEquals(1, renamedT.inputSpecs().size());
        assertEquals("i1/in", renamedT.inputSpecs().get(0).place().name());
    }

    @Test
    void renameNet_clearsRemapMaps_onEntry() {
        Place<String> p1 = p("p1", String.class);
        Transition t = Transition.builder("t").read(p1).build();
        PetriNet body = PetriNet.builder("B").transition(t).build();

        var placeRemap = new HashMap<Place<?>, Place<?>>();
        placeRemap.put(p("garbage", String.class), p("more-garbage", String.class));
        var transitionRemap = new HashMap<Transition, Transition>();
        transitionRemap.put(t, t); // garbage stale entry

        SubnetRewriter.renameNet(body, "x", placeRemap, transitionRemap);

        // Garbage entries must be wiped out — only the freshly-renamed entries remain.
        assertFalse(placeRemap.containsKey(p("garbage", String.class)),
            "renameNet must clear the placeRemap on entry");
        assertEquals(1, placeRemap.size());
        assertTrue(placeRemap.containsKey(p1));

        assertEquals(1, transitionRemap.size());
        // The fresh entry has the renamed transition, not the original.
        assertEquals("x/t", transitionRemap.get(t).name());
    }

    @Test
    void rewriteTransition_preservesTimingPriorityActionByReference() {
        Place<String> in = p("in", String.class);
        // Use a non-passthrough action so we can compare references.
        org.libpetri.core.TransitionAction sharedAction =
            ctx -> java.util.concurrent.CompletableFuture.completedFuture(null);

        Transition orig = Transition.builder("t")
            .inputs(Arc.In.one(in))
            .timing(org.libpetri.core.Timing.delayed(Duration.ofMillis(50)))
            .priority(7)
            .action(sharedAction)
            .build();

        var m = remap(in, "z");
        var rewritten = SubnetRewriter.rewriteTransition(orig, "z", m);

        assertEquals("z/t", rewritten.name());
        assertEquals(7, rewritten.priority());
        assertSame(sharedAction, rewritten.action(), "MOD-030: action shared by reference");
        assertEquals(orig.timing(), rewritten.timing());
    }

    // ============================================================
    //  Convenience factories
    // ============================================================

    @Test
    void newPlaceRemap_returnsEmptyMutableMap() {
        Map<Place<?>, Place<?>> m = SubnetRewriter.newPlaceRemap();
        assertTrue(m.isEmpty());
        m.put(p("x", String.class), p("y", String.class));
        assertEquals(1, m.size());
    }

    @Test
    void newTransitionRemap_returnsEmptyMutableMap() {
        Map<Transition, Transition> m = SubnetRewriter.newTransitionRemap();
        assertTrue(m.isEmpty());
        Transition t = Transition.builder("t").read(p("p", String.class)).build();
        m.put(t, t);
        assertEquals(1, m.size());
    }

}

File: typescript/tests/core/internal/subnet-rewriter.test.ts
import { describe, it, expect } from 'vitest';
import { place } from '../../../src/core/place.js';
import type { Place } from '../../../src/core/place.js';
import { Transition } from '../../../src/core/transition.js';
import { PetriNet } from '../../../src/core/petri-net.js';
import {
  one,
  exactly,
  all,
  atLeast,
  type In,
} from '../../../src/core/in.js';
import {
  outPlace,
  and,
  xor,
  timeout,
  forwardInput,
  type Out,
} from '../../../src/core/out.js';
import {
  renamePlace,
  renameNet,
  rewriteTransition,
  substitutePlaces,
  rewriteIn,
  rewriteOut,
  rewriteInhibitor,
  rewriteRead,
  rewriteReset,
} from '../../../src/core/internal/subnet-rewriter.js';

/**
 * Unit tests for the package-internal subnet rewriter. The rewriter is the
 * structural rewrite primitive used by `SubnetDef.instantiate` and (later)
 * `PetriNetBuilder.compose(...)`. These tests pin the per-arc-variant
 * rewrite contract and the `renameNet` end-to-end rename pass.
 */
describe('SubnetRewriter', () => {
  // ============================================================
  //  renamePlace — basic shape
  // ============================================================

  describe('renamePlace', () => {
    it('prefixes the name with prefix + "/"', () => {
      const orig = place<string>('p');
      const renamed = renamePlace(orig, 'pre');
      expect(renamed.name).toBe('pre/p');
    });

    it('returns a fresh Place reference (does not mutate input)', () => {
      const orig = place<string>('p');
      const renamed = renamePlace(orig, 'pre');
      expect(renamed).not.toBe(orig);
      expect(orig.name).toBe('p');
    });
  });

  // ============================================================
  //  resolve / partial-remap behavior — places not in remap are kept
  // ============================================================

  describe('partial-remap (places not in remap pass through)', () => {
    it('rewriteIn with empty remap returns identical-shaped In with same place', () => {
      const p = place<number>('p');
      const remap = new Map<string, Place<unknown>>();
      const rewritten = rewriteIn(one(p), remap) as { type: 'one'; place: Place<unknown> };
      expect(rewritten.type).toBe('one');
      expect(rewritten.place).toBe(p);
    });

    it('rewriteOut with empty remap on a Place leaf preserves the place', () => {
      const p = place<string>('out');
      const remap = new Map<string, Place<unknown>>();
      const rewritten = rewriteOut(outPlace(p), remap) as { type: 'place'; place: Place<unknown> };
      expect(rewritten.type).toBe('place');
      expect(rewritten.place).toBe(p);
    });

    it('rewriteInhibitor with partial remap substitutes only known places', () => {
      const p = place<string>('p');
      const q = place<string>('q');
      const renamedP = place<string>('pre/p');
      const remap = new Map<string, Place<unknown>>([['p', renamedP as Place<unknown>]]);

      // p is in remap, q is not.
      const rewrittenP = rewriteInhibitor({ type: 'inhibitor', place: p }, remap);
      const rewrittenQ = rewriteInhibitor({ type: 'inhibitor', place: q }, remap);

      expect(rewrittenP.place).toBe(renamedP);
      expect(rewrittenQ.place).toBe(q);
    });
  });

  // ============================================================
  //  rewriteIn — exhaustive over In discriminator
  // ============================================================

  describe('rewriteIn — all variants', () => {
    function makeRemap(orig: Place<unknown>, prefix: string): Map<string, Place<unknown>> {
      return new Map([[orig.name, place<unknown>(prefix + '/' + orig.name)]]);
    }

    it('one: substitutes place, preserves type discriminator', () => {
      const p = place<number>('p');
      const remap = makeRemap(p as Place<unknown>, 'pre');
      const rewritten = rewriteIn(one(p), remap);
      expect(rewritten.type).toBe('one');
      expect(rewritten.place.name).toBe('pre/p');
    });

    it('exactly: preserves count and substitutes place', () => {
      const p = place<number>('p');
      const remap = makeRemap(p as Place<unknown>, 'pre');
      const rewritten = rewriteIn(exactly(3, p), remap) as Extract<In, { type: 'exactly' }>;
      expect(rewritten.type).toBe('exactly');
      expect(rewritten.count).toBe(3);
      expect(rewritten.place.name).toBe('pre/p');
    });

    it('all: substitutes place', () => {
      const p = place<number>('p');
      const remap = makeRemap(p as Place<unknown>, 'pre');
      const rewritten = rewriteIn(all(p), remap) as Extract<In, { type: 'all' }>;
      expect(rewritten.type).toBe('all');
      expect(rewritten.place.name).toBe('pre/p');
    });

    it('at-least: preserves minimum and substitutes place', () => {
      const p = place<number>('p');
      const remap = makeRemap(p as Place<unknown>, 'pre');
      const rewritten = rewriteIn(atLeast(2, p), remap) as Extract<In, { type: 'at-least' }>;
      expect(rewritten.type).toBe('at-least');
      expect(rewritten.minimum).toBe(2);
      expect(rewritten.place.name).toBe('pre/p');
    });
  });

  // ============================================================
  //  rewriteOut — exhaustive over Out discriminator + recursion
  // ============================================================

  describe('rewriteOut — all variants', () => {
    function makeRemap(prefix: string, ...originals: Place<unknown>[]): Map<string, Place<unknown>> {
      const m = new Map<string, Place<unknown>>();
      for (const p of originals) m.set(p.name, place<unknown>(prefix + '/' + p.name));
      return m;
    }

    it('place: substitutes place', () => {
      const p = place<string>('p');
      const remap = makeRemap('pre', p as Place<unknown>);
      const rewritten = rewriteOut(outPlace(p), remap) as Extract<Out, { type: 'place' }>;
      expect(rewritten.type).toBe('place');
      expect(rewritten.place.name).toBe('pre/p');
    });

    it('forward-input: substitutes both from and to', () => {
      const from = place<string>('from');
      const to = place<string>('to');
      const remap = makeRemap('pre', from as Place<unknown>, to as Place<unknown>);
      const rewritten = rewriteOut(forwardInput(from, to), remap) as Extract<Out, { type: 'forward-input' }>;
      expect(rewritten.type).toBe('forward-input');
      expect(rewritten.from.name).toBe('pre/from');
      expect(rewritten.to.name).toBe('pre/to');
    });

    it('and: substitutes children in order', () => {
      const a = place<string>('a');
      const b = place<string>('b');
      const remap = makeRemap('pre', a as Place<unknown>, b as Place<unknown>);
      const original = and(outPlace(a), outPlace(b));
      const rewritten = rewriteOut(original, remap) as Extract<Out, { type: 'and' }>;
      expect(rewritten.type).toBe('and');
      expect(rewritten.children).toHaveLength(2);
      expect((rewritten.children[0] as Extract<Out, { type: 'place' }>).place.name).toBe('pre/a');
      expect((rewritten.children[1] as Extract<Out, { type: 'place' }>).place.name).toBe('pre/b');
    });

    it('xor: substitutes children in order', () => {
      const a = place<string>('a');
      const b = place<string>('b');
      const remap = makeRemap('pre', a as Place<unknown>, b as Place<unknown>);
      const original = xor(outPlace(a), outPlace(b));
      const rewritten = rewriteOut(original, remap) as Extract<Out, { type: 'xor' }>;
      expect(rewritten.type).toBe('xor');
      expect(rewritten.children).toHaveLength(2);
      expect((rewritten.children[0] as Extract<Out, { type: 'place' }>).place.name).toBe('pre/a');
      expect((rewritten.children[1] as Extract<Out, { type: 'place' }>).place.name).toBe('pre/b');
    });

    it('timeout: preserves afterMs, recurses into child', () => {
      const p = place<string>('p');
      const remap = makeRemap('pre', p as Place<unknown>);
      const rewritten = rewriteOut(timeout(150, outPlace(p)), remap) as Extract<Out, { type: 'timeout' }>;
      expect(rewritten.type).toBe('timeout');
      expect(rewritten.afterMs).toBe(150);
      expect((rewritten.child as Extract<Out, { type: 'place' }>).place.name).toBe('pre/p');
    });

    it('and(xor(...), and(...)): nested recursion through both branches', () => {
      const a = place<string>('a');
      const b = place<string>('b');
      const c = place<string>('c');
      const d = place<string>('d');
      const remap = makeRemap(
        'pre',
        a as Place<unknown>,
        b as Place<unknown>,
        c as Place<unknown>,
        d as Place<unknown>,
      );

      const nested = and(xor(outPlace(a), outPlace(b)), and(outPlace(c), outPlace(d)));
      const rewritten = rewriteOut(nested, remap) as Extract<Out, { type: 'and' }>;

      expect(rewritten.type).toBe('and');
      const xorBranch = rewritten.children[0] as Extract<Out, { type: 'xor' }>;
      expect(xorBranch.type).toBe('xor');
      expect((xorBranch.children[0] as Extract<Out, { type: 'place' }>).place.name).toBe('pre/a');
      expect((xorBranch.children[1] as Extract<Out, { type: 'place' }>).place.name).toBe('pre/b');

      const andBranch = rewritten.children[1] as Extract<Out, { type: 'and' }>;
      expect(andBranch.type).toBe('and');
      expect((andBranch.children[0] as Extract<Out, { type: 'place' }>).place.name).toBe('pre/c');
      expect((andBranch.children[1] as Extract<Out, { type: 'place' }>).place.name).toBe('pre/d');
    });

    it('timeout(and(forward-input)): triple-nested recursion', () => {
      const from = place<string>('from');
      const to = place<string>('to');
      const remap = new Map<string, Place<unknown>>([
        ['from', place<unknown>('pre/from')],
        ['to', place<unknown>('pre/to')],
      ]);

      const out: Out = timeout(50, and(forwardInput(from, to)));
      const rewritten = rewriteOut(out, remap) as Extract<Out, { type: 'timeout' }>;
      expect(rewritten.type).toBe('timeout');
      expect(rewritten.afterMs).toBe(50);
      const andChild = rewritten.child as Extract<Out, { type: 'and' }>;
      expect(andChild.type).toBe('and');
      const fwd = andChild.children[0] as Extract<Out, { type: 'forward-input' }>;
      expect(fwd.type).toBe('forward-input');
      expect(fwd.from.name).toBe('pre/from');
      expect(fwd.to.name).toBe('pre/to');
    });
  });

  // ============================================================
  //  rewriteInhibitor / rewriteRead / rewriteReset
  // ============================================================

  describe('rewriteInhibitor / rewriteRead / rewriteReset', () => {
    it('inhibitor: substitutes place', () => {
      const p = place<string>('p');
      const renamed = place<unknown>('pre/p');
      const remap = new Map<string, Place<unknown>>([['p', renamed]]);
      const rewritten = rewriteInhibitor({ type: 'inhibitor', place: p }, remap);
      expect(rewritten.type).toBe('inhibitor');
      expect(rewritten.place).toBe(renamed);
    });

    it('read: substitutes place', () => {
      const p = place<string>('p');
      const renamed = place<unknown>('pre/p');
      const remap = new Map<string, Place<unknown>>([['p', renamed]]);
      const rewritten = rewriteRead({ type: 'read', place: p }, remap);
      expect(rewritten.type).toBe('read');
      expect(rewritten.place).toBe(renamed);
    });

    it('reset: substitutes place', () => {
      const p = place<string>('p');
      const renamed = place<unknown>('pre/p');
      const remap = new Map<string, Place<unknown>>([['p', renamed]]);
      const rewritten = rewriteReset({ type: 'reset', place: p }, remap);
      expect(rewritten.type).toBe('reset');
      expect(rewritten.place).toBe(renamed);
    });
  });

  // ============================================================
  //  rewriteTransition — name prefix + arc rewrite + action by-ref
  // ============================================================

  describe('rewriteTransition', () => {
    it('prefixes name and rewrites arcs; preserves action by reference', () => {
      const p = place<string>('p');
      const q = place<string>('q');
      const action = async () => {};
      const t = Transition.builder('t')
        .inputs(one(p))
        .outputs(outPlace(q))
        .priority(2)
        .action(action)
        .build();

      const remap = new Map<string, Place<unknown>>([
        ['p', place<unknown>('pre/p')],
        ['q', place<unknown>('pre/q')],
      ]);
      const renamed = rewriteTransition(t, 'pre', remap);

      expect(renamed.name).toBe('pre/t');
      expect(renamed.priority).toBe(2);
      expect(renamed.action).toBe(action);
      expect(renamed.inputSpecs).toHaveLength(1);
      expect((renamed.inputSpecs[0] as Extract<In, { type: 'one' }>).place.name).toBe('pre/p');
      expect(renamed.outputSpec).not.toBeNull();
      expect((renamed.outputSpec as Extract<Out, { type: 'place' }>).place.name).toBe('pre/q');
    });

    it('preserves inhibitor / read / reset arcs', () => {
      const p = place<string>('p');
      const inh = place<string>('inh');
      const rd = place<string>('rd');
      const rs = place<string>('rs');

      const t = Transition.builder('t')
        .read(p)
        .inhibitor(inh)
        .read(rd)
        .reset(rs)
        .build();

      const remap = new Map<string, Place<unknown>>([
        ['p', place<unknown>('pre/p')],
        ['inh', place<unknown>('pre/inh')],
        ['rd', place<unknown>('pre/rd')],
        ['rs', place<unknown>('pre/rs')],
      ]);
      const renamed = rewriteTransition(t, 'pre', remap);

      expect(renamed.inhibitors).toHaveLength(1);
      expect(renamed.inhibitors[0]!.place.name).toBe('pre/inh');
      expect(renamed.reads).toHaveLength(2);
      expect(renamed.reads.map((r) => r.place.name).sort()).toEqual(['pre/p', 'pre/rd']);
      expect(renamed.resets).toHaveLength(1);
      expect(renamed.resets[0]!.place.name).toBe('pre/rs');
    });
  });

  // ============================================================
  //  substitutePlaces — same name, place substitution only
  // ============================================================

  describe('substitutePlaces', () => {
    it('keeps the original name; substitutes places only', () => {
      const p = place<string>('p');
      const t = Transition.builder('keep_my_name').inputs(one(p)).build();
      const remap = new Map<string, Place<unknown>>([
        ['p', place<unknown>('caller/p')],
      ]);

      const rewritten = substitutePlaces(t, remap);
      expect(rewritten.name).toBe('keep_my_name');
      expect((rewritten.inputSpecs[0] as Extract<In, { type: 'one' }>).place.name).toBe('caller/p');
    });

    it('partial remap: places not in remap pass through unchanged', () => {
      const p = place<string>('p');
      const q = place<string>('q');
      const t = Transition.builder('t').inputs(one(p), one(q)).build();
      const remap = new Map<string, Place<unknown>>([
        ['p', place<unknown>('caller/p')],
      ]);

      const rewritten = substitutePlaces(t, remap);
      expect(rewritten.inputSpecs).toHaveLength(2);
      const placeNames = rewritten.inputSpecs.map(
        (s) => (s as Extract<In, { type: 'one' }>).place.name,
      );
      expect(placeNames).toContain('caller/p');
      expect(placeNames).toContain('q');
    });
  });

  // ============================================================
  //  renameNet — end-to-end rename pass + remap-side-effect contract
  // ============================================================

  describe('renameNet', () => {
    it('renames every place and transition; fills both remaps; preserves stranded places', () => {
      const a = place<string>('a');
      const b = place<string>('b');
      const stranded = place<string>('stranded');
      const t = Transition.builder('t').inputs(one(a)).outputs(outPlace(b)).build();
      const body = PetriNet.builder('body').place(stranded).transition(t).build();

      const placeRemap = new Map<string, Place<unknown>>();
      const transitionRemap = new Map<string, Transition>();
      const renamed = renameNet(body, 'pre', placeRemap, transitionRemap);

      // Net name is prefixed.
      expect(renamed.name).toBe('pre/body');

      // All places renamed AND stranded place is preserved.
      const renamedPlaceNames = [...renamed.places].map((p) => p.name).sort();
      expect(renamedPlaceNames).toEqual(['pre/a', 'pre/b', 'pre/stranded']);

      // Transition renamed.
      expect(renamed.transitions.size).toBe(1);
      const [renamedT] = [...renamed.transitions];
      expect(renamedT!.name).toBe('pre/t');

      // Remaps populated.
      expect(placeRemap.get('a')?.name).toBe('pre/a');
      expect(placeRemap.get('b')?.name).toBe('pre/b');
      expect(placeRemap.get('stranded')?.name).toBe('pre/stranded');
      expect(transitionRemap.get('t')).toBe(renamedT);
    });

    it('clears caller-supplied maps before populating', () => {
      const p = place<string>('p');
      const t = Transition.builder('t').read(p).build();
      const body = PetriNet.builder('body').transition(t).build();

      const placeRemap = new Map<string, Place<unknown>>([
        ['stale', place<unknown>('stale_value')],
      ]);
      const transitionRemap = new Map<string, Transition>([
        ['stale_t', t],
      ]);

      renameNet(body, 'pre', placeRemap, transitionRemap);

      expect(placeRemap.has('stale')).toBe(false);
      expect(transitionRemap.has('stale_t')).toBe(false);
      expect(placeRemap.get('p')?.name).toBe('pre/p');
    });
  });
});
